File: Fore-C_v2/codes/Grid_covariates_ocean_color_dynamic.R
# Pull ocean color  co-variates to grid -----------------------------

# load libraries
library(ncdf4)
library(raster)
library(tidyverse)

# source custom function
source("./codes/custom_functions/fun_extract_oc_data_from_netcdf.R")

# list oc co-variates ----------------------------------------------
# source("./codes/Final_covariates_by_disease_and_region.R")
# 
# all_covars <- c(ga_gbr_vars
#                 , ga_pac_vars
#                 , ws_gbr_vars
#                 , ws_pac_acr_vars)
# 
# all_covars <- unique(all_covars)
# 
# oc_covars <- all_covars[grep('Kd', all_covars)]

# load data ----------------------------------------------------------
# load SST data to get matching dates 
load("../compiled_data/grid_covariate_data/grid_with_sst_metrics.RData")
oc_dates <- unique(reef_grid_sst$Date)

# load ocean color data
oc_filepath <- "../raw_data/covariate_data/ocean_color/long_term_metrics_20211121.nc"
oc <- nc_open(oc_filepath)

# load reef grid
load("../compiled_data/spatial_data/grid.RData")

# Make vector of latitudes and longitudes (same for all data) ------------------
lons <- oc$dim$lon$vals
lats <- oc$dim$lat$vals

# Extract seasonal ocean color data by pixel -----------------------------------

# turn list of dates in list of Julian days
julian_days <- as.numeric(strftime(oc_dates, 
                                   format = "%j")) # 14 weeks of predictions

# create empty data frames for data
TW_median <- data.frame()
TW_90th <- data.frame()

# loop through metrics and dates; this takes ~6 minutes
oc_metrics <- c("three_week_kd490_median", 
                "three_week_kd490_90th")

oc_metrics_names <- c("Three_Week_Kd_Median",
                      "Three_Week_Kd_90th")

for(i in 1:length(oc_metrics)){
  for(j in 1:length(julian_days)){
    x <- ncvar_get(oc, 
                   varid = oc_metrics[i], 
                   start = c(1, 1, julian_days[j]),
                   count = c(7200, 3600, 1))
    
    x2 <- extract_oc_data(variableMatrix = x,
                          reefgrid = reefsDF)
    
    tmp_df <- reefsDF
    tmp_df$Date <- oc_dates[j]
    tmp_df[, oc_metrics_names[i]] <- x2
    if(oc_metrics_names[i] == "Three_Week_Kd_Median"){
      TW_median <- rbind(TW_median, tmp_df)
    } else {
      TW_90th <- rbind(TW_90th, tmp_df)
    }
  }
}

tw_metrics <- TW_median %>%
  left_join(TW_90th, 
            by = c("Longitude", 
                   "Latitude", 
                   "Region", 
                   "ID", 
                   "Date"))

tw_metrics$Three_Week_Kd_Variability <- tw_metrics$Three_Week_Kd_90th - tw_metrics$Three_Week_Kd_Median

reef_grid_tw_oc <- tw_metrics[, c("Longitude",
                             "Latitude",
                             "Region",
                             "ID",
                             "Date",
                             "Three_Week_Kd_Median",
                             "Three_Week_Kd_Variability")]

save(reef_grid_tw_oc, file = "../compiled_data/grid_covariate_data/grid_with_three_week_oc_metrics.RData")

File: Fore-C_v2/codes/Grid_covariates_ocean_color_dynamic.py
# -*- coding: utf-8 -*-
"""
Created on Thu Apr 28 09:18:45 2022

@author: jamie
"""

import netCDF4 as nc
import pandas as pd
import numpy as np

# load reef grid
reefsDF = pd.read_csv('../compiled_data/spatial_data/grid.csv')

# load SST data to get matching dates
oc_dates = pd.read_csv('../compiled_data/grid_covariate_data/grid_with_sst_metrics.csv')
oc_dates = oc_dates.Date.unique()
oc_dates = pd.to_datetime(oc_dates, format='%Y-%m-%d')
oc_doy = oc_dates.strftime('%j')

# load ocean color data
oc_filepath = '../raw_data/covariate_data/ocean_color/'
oc_long_term = nc.Dataset(oc_filepath + 'long_term_metrics_20211121.nc')
oc_pixel_id = nc.Dataset(oc_filepath + 'reef_grid_pixel_id.nc')

# create an array of ids
ids = oc_pixel_id.variables['pixel_id'][:]
ids_loc = np.argwhere(ids != -999.0)

ids_with_values = ids[ids_loc[:,0], ids_loc[:,1]]

# create array of Kd(490) values
tw_kd490_median = oc_long_term.variables['three_week_kd490_median']
tw_kd490_90th = oc_long_term.variables['three_week_kd490_90th']

# loop through all days and extract values for reefs pixel
# this loop takes about 1-4 minutes on standard laptop
oc_metrics = pd.DataFrame()

for i in range(len(oc_doy)):
    # subset to correct day
    median = tw_kd490_median[oc_doy[i], :, :]
    kd90th = tw_kd490_90th[oc_doy[i], :, :]
    # get values for specified pixels
    median_values = median[ids_loc[:,0], ids_loc[:,1]]
    kd90th_values = kd90th[ids_loc[:,0], ids_loc[:,1]]
    # get values from masked arrays
    # median_values = median_values[median_values.mask == False]
    # kd90th_values = kd90th_values[kd90th_values.mask == False]
    # create temporary dataframe
    tmp_df = pd.DataFrame()
    tmp_df['ID'] = ids_with_values
    tmp_df['Date'] = oc_dates[i]
    tmp_df['Three_Week_Kd_Median'] = median_values
    tmp_df['Three_week_kd490_90th'] = kd90th_values
    # add to ocean color dataset
    oc_metrics = oc_metrics.append(tmp_df)

# add new column for variability
oc_metrics['Three_Week_Kd_Variability'] = oc_metrics['Three_week_kd490_90th'] - oc_metrics['Three_Week_Kd_Median']

# change pixel ID from float to integer
oc_metrics['ID'] = oc_metrics['ID'].astype(int)

# fill NaN with pixel-specific mean (across all dates)
oc_metrics['Three_Week_Kd_Median'] = oc_metrics['Three_Week_Kd_Median'].fillna(oc_metrics.groupby('ID')['Three_Week_Kd_Median'].transform('mean'))
oc_metrics['Three_Week_Kd_Variability'] = oc_metrics['Three_Week_Kd_Variability'].fillna(oc_metrics.groupby('ID')['Three_Week_Kd_Variability'].transform('mean'))

# fill remaining NaN with date-specific mean (across all pixels)
oc_metrics['Three_Week_Kd_Median'] = oc_metrics['Three_Week_Kd_Median'].fillna(oc_metrics.groupby('Date')['Three_Week_Kd_Median'].transform('mean'))
oc_metrics['Three_Week_Kd_Variability'] = oc_metrics['Three_Week_Kd_Variability'].fillna(oc_metrics.groupby('Date')['Three_Week_Kd_Variability'].transform('mean'))

# save
oc_metrics.to_csv('../compiled_data/grid_covariate_data/grid_with_three_week_oc_metrics.csv', index = False)
    
